lib/SILOptimizer/Transforms/SemanticARCOpts.cpp

```diff
@@ -41,6 +41,92 @@ STATISTIC(NumEliminatedInsts, "number of removed instructions");
 STATISTIC(NumLoadCopyConvertedToLoadBorrow,
           "number of load_copy converted to load_borrow");
 
+//===----------------------------------------------------------------------===//
+//                           Ownership Phi Operand
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+/// The operand of a "phi" in the induced ownership graph of a def-use graph.
+///
+/// Some examples: br, struct, tuple.
+class OwnershipPhiOperand {
+public:
+  enum Kind {
+    Branch,
+  };
+
+private:
+  Operand *op;
+
+  OwnershipPhiOperand(Operand *op) : op(op) {}
+
+public:
+  static Optional<OwnershipPhiOperand> get(const Operand *op) {
+    switch (op->getUser()->getKind()) {
+    case SILInstructionKind::BranchInst:
+      return {{const_cast<Operand *>(op)}};
+    default:
+      return None;
+    }
+  }
+
+  Kind getKind() const {
+    switch (op->getUser()->getKind()) {
+    case SILInstructionKind::BranchInst:
+      return Kind::Branch;
+    default:
+      llvm_unreachable("unhandled case?!");
+    }
+  }
+
+  Operand *getOperand() const { return op; }
+  SILValue getValue() const { return op->get(); }
+  SILType getType() const { return op->get()->getType(); }
+
+  unsigned getOperandNumber() const { return op->getOperandNumber(); }
+
+  void markUndef() & {
+    op->set(SILUndef::get(getType(), *op->getUser()->getFunction()));
+  }
+
+  SILInstruction *getInst() const { return op->getUser(); }
+
+  /// Return true if this phi consumes a borrow.
+  ///
+  /// If so, we may need to insert an extra begin_borrow to balance the +1 when
+  /// converting owned ownership phis to guaranteed ownership phis.
+  bool isGuaranteedConsuming() const {
+    switch (getKind()) {
+    case Kind::Branch:
+      return true;
+    }
+  }
+
+  bool operator<(const OwnershipPhiOperand &other) const {
+    return op < other.op;
+  }
+
+  bool operator==(const OwnershipPhiOperand &other) const {
+    return op == other.op;
+  }
+
+  bool visitResults(function_ref<bool(SILValue)> visitor) const {
+    switch (getKind()) {
+    case Kind::Branch: {
+      auto *br = cast<BranchInst>(getInst());
+      unsigned opNum = getOperandNumber();
+      return llvm::all_of(
```

Comment:
One thing to be aware of is that when I add instructions like struct/tuple here, I am going to be just returning getInst() essentially.